server/schedule/namespace_checker.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedule
+
+import (
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/namespace"
+)
+
+// NamespaceChecker ensures region to go to the right place.
+type NamespaceChecker struct {
+	opt        Options
+	cluster    Cluster
+	filters    []Filter
+	classifier namespace.Classifier
+}
+
+// NewNamespaceChecker creates a namespace checker.
+func NewNamespaceChecker(opt Options, cluster Cluster, classifier namespace.Classifier) *NamespaceChecker {
+	filters := []Filter{
+		NewHealthFilter(opt),
+		NewSnapshotCountFilter(opt),
+	}
+
+	return &NamespaceChecker{
+		opt:        opt,
+		cluster:    cluster,
+		filters:    filters,
+		classifier: classifier,
+	}
+}
+
+// Check verifies a region's namespace, creating an Operator if need.
+func (n *NamespaceChecker) Check(region *core.RegionInfo) *Operator {
+	// fail-fast if there is only ONE namespace
+	if n.classifier == nil || len(n.classifier.GetAllNamespaces()) == 1 {
+		return nil
+	}
+	for _, peer := range region.GetPeers() {
+		newPeer := n.SelectBestPeerToRelocate(region, n.filters...)
```

Comment:
I think we need to check if `peer` is located on a store that is already belongs to the right namespace.

Reply:
change logic which checking the `peer` location first